### PR TITLE
A0-581: Smarternet

### PR DESCRIPTION
**Pull request by h4nsu**
A candidate for Smartnet 2.0 i.e. a new version of node which supports the stable release of ink 3.0.0
This PR is based on #359.

NOTE: This PR is meant to be merged into `smartnet` branch, not into `main`. However, `smartnet` branch is quite old and picking it as base would show several monts of `main` evolution in diff rendering code review undoable. Hence the `main` branch is picked as a temporary base for the time of the review.

### CODE REVIEW

File: bin/runtime/src/lib.rs [hunk 1]
```diff
@@ -789,4 +834,49 @@ impl_runtime_apis! {
                 .collect::<Result<Vec<AlephId>, AlephApiError>>()
         }
     }
+
+    impl pallet_contracts_rpc_runtime_api::ContractsApi<Block, AccountId, Balance, BlockNumber, Hash,> for Runtime {
```
Comment (kostekIV) [inline bin/runtime/src/lib.rs:838]: opinionated nit: `,` at the end looks strange to me :|

File: bin/runtime/src/lib.rs [hunk 2]
```diff
@@ -607,6 +611,46 @@ impl pallet_utility::Config for Runtime {
...
+    type CallFilter = Nothing;
+    type DepositPerItem = DepositPerItem;
+    type DepositPerByte = DepositPerByte;
+    type WeightPrice = pallet_transaction_payment::Pallet<Self>;
+    type WeightInfo = pallet_contracts::weights::SubstrateWeight<Self>;
+    type ChainExtension = ();
+    type DeletionQueueDepth = DeletionQueueDepth;
+    type DeletionWeightLimit = DeletionWeightLimit;
+    type Schedule = Schedule;
+    type CallStack = [pallet_contracts::Frame<Self>; 31];
```
Comment (kostekIV) [inline bin/runtime/src/lib.rs:650]: why 31 and not 32? :O

File: bin/runtime/src/lib.rs [hunk 3]
```diff
@@ -607,6 +611,46 @@ impl pallet_utility::Config for Runtime {
     type PalletsOrigin = OriginCaller;
 }
 
+parameter_types! {
+    // The following weights are minimal to allow Smartnet users do whatever they want.
+    pub const DepositPerItem: Balance = 1_000_000_000;  //  1 milli-SZERO per item
+    pub const DepositPerByte: Balance = 1_000_000;      // ~1 milli-SZERO per kB storage
```
Comment (kostekIV) [inline bin/runtime/src/lib.rs:617]: use here already defined `TOKEN` const: `TOKEN / 1_000` etc

**Review by kostekIV — APPROVED** [review]
LGTM

File: bin/runtime/src/lib.rs (hunk 2 above)
Comment (h4nsu) [reply bin/runtime/src/lib.rs:650]: I copied it from a tempate node `substrate-contracts-node` ^^

File: bin/runtime/src/lib.rs (hunk 3 above)
Comment (h4nsu) [reply bin/runtime/src/lib.rs]: Done

**Review by fbielejec — APPROVED** [review]
Looks good, just don;t forget to merge it to the smartnet I guess

File: bin/runtime/src/lib.rs (hunk 2 above)
Comment (pmikolajczyk41) [reply bin/runtime/src/lib.rs:650]: 1 is added implicitly, i.e. the initial stack frame; thus there are 32 possible levels

File: bin/runtime/src/lib.rs [hunk 4]
```diff
@@ -607,6 +611,46 @@ impl pallet_utility::Config for Runtime {
...
+    // The lazy deletion runs inside on_initialize.
+    pub DeletionWeightLimit: Weight = Perbill::from_percent(10) * BlockWeights::get().max_block;
+    // The weight needed for decoding the queue should be less or equal than a fifth
+    // of the overall weight dedicated to the lazy deletion.
+    pub DeletionQueueDepth: u32 = ((DeletionWeightLimit::get() / (
+            <Runtime as pallet_contracts::Config>::WeightInfo::on_initialize_per_queue_item(1) -
+            <Runtime as pallet_contracts::Config>::WeightInfo::on_initialize_per_queue_item(0)
+        )) / 5) as u32;
```
[the remark this answers is not in the document]
Comment (pmikolajczyk41) [reply bin/runtime/src/lib.rs]: My only concern is that users may ask and the only answer we could provide is: somebody at Parity wrote that :)

**Review by pmikolajczyk41 — APPROVED** [review]

**Pull request merged**